working/macros/feature-specification.md

```diff
@@ -1300,28 +1300,68 @@ given library. This will be allowed through the library introspection class,
 which is available from the introspection APIs on all declarations via a
 `library` getter.
 
-**TODO**: Fully define the library introspection API for each phase.
-
 ### API versioning
 
-**TODO**: Finalize the approach here (#1934).
-
-It is possible that future language changes would require a breaking change to
-an existing imperative macro API. For instance you could consider what would
-happen if we added multiple return values from functions. That would
-necessitate a change to many APIs so that they would support multiple return
-types instead of a single one.
-
-#### Proposal: Ship macro APIs as a Pub package
-
-Likely, this package would only export directly an existing `dart:` uri, but
-it would be able to be versioned like a public package, including tight sdk
-constraints (likely on minor version ranges). This would work similar to the
-`dart:_internal` library.
-
-This approach would involve more work on our end (to release this package with
-each dart release). But it would help keep users on the rails, and give us a
-lot of flexibility with the API going forward.
+It is expected that future language changes will require breaking changes to the
+macro APIs. For instance you could consider what would happen if we added
+multiple return values from functions. That would necessitate a change to many
+APIs so that they would support multiple return types instead of a single one.
+
+#### Design goals
+
+* Enable us to make API changes when needed.
+* Minimize churn for macro authors - most SDK versions will not be breaking and
+  it would be ideal to not ask packages to have super tight SDK constraints.
+* Give early and actionable error messages to consumers of macros when a macro
+  they are using does not support a new language feature.
+* Enable as much forwards/backwards compatibility as possible.
+
+#### Solution: Ship macro APIs as a Pub package
+
+The implementation of this package will always come from the SDK, through a
+`dart:_macros` library, which will be exported by this package. The
+`dart:_macros` library will be blocked from being imported or exported by any
+library other than other `dart:` libraries and this package.
+
+We use a `dart:` library here to ensure that the binaries shipped with the SDK
+are compiled with exactly the same version of the API that macros are compiled
+with. This ensures the communication protocol between macros and the SDK
+binaries are compatible.
+
+This package will have tight upper bound SDK constraints, and will need to be
+released any time a new minor release is published. It is expected that most of
+these releases would be patch releases, which only increase the max SDK
+constraint.
```

Comment:
This is a little opaque if you are not steeped in package versioning. Maybe walk through two scenarios:

* When a new version of the Dart SDK is released which does not have a breaking change in the macro API, then we will release a patch version of the macro package which simply expands the SDK upper bound to include that version. Since this is just a patch release, all existing packages that depend on the macro package support it already, and no work is required on macro authors' part to work with the new Dart SDK.

* When a new version of the Dart SDK does include a breaking change to the macro API, then we will release a new version of the macro package with an increased *major* version number. By default, existing packages containing macros will not accept that version of the macro package and thus will not work with the new Dart SDK.

    Authors of packages containing macros will need to test to see if their macro is compatible with the latest macro API. If so, they can ship a new patch version of their package with a constraint on the macro package that includes the new major version as well as the previous major version it's already known to work with. If their package is broken by the macro API change, then, they will fix their macro and ship a new version of their package with a dependency on the macro package that *only* allows the latest major versions.

(Walking through an example with concrete version numbers and package names would be even better, but you get the idea.)

Reply:
Updated